File: 322-coin-change/coin-change.swift
// You are given coins of different denominations and a total amount of money amount. Write a function to compute the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.
//
// Example 1:
//
//
// Input: coins = [1, 2, 5], amount = 11
// Output: 3 
// Explanation: 11 = 5 + 5 + 1
//
// Example 2:
//
//
// Input: coins = [2], amount = 3
// Output: -1
//
//
// Note:
// You may assume that you have an infinite number of each kind of coin.
//


class Solution {
    func coinChange(_ coins: [Int], _ amount: Int) -> Int {
        if amount == 0 {
            return 0
        }
     
        var coinNums = Array(repeating: -1, count: amount + 1)
        for coin in coins {
            if coin < amount {
                coinNums[coin] = 1
            }
            else if coin == amount {
                return 1
            }
        }
        
        var numsQueue = coins
        while !numsQueue.isEmpty {
            let val = numsQueue.first!
            numsQueue.remove(at: 0)
            
            for coin in coins {
                let newValue = coin + val
                if newValue == amount {
                    return coinNums[val] + 1
                }
                else if newValue < amount {
                    if coinNums[newValue] == -1 {
                        if coinNums[val] != -1 {
                            coinNums[newValue] = coinNums[val] + 1
                            numsQueue.append(newValue)
                        }
                    }
                }
            }
        }
        
        return -1
    }
}

File: 322-coin-change/coin-change.cpp
// You are given coins of different denominations and a total amount of money amount. Write a function to compute the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.
//
// Example 1:
//
//
// Input: coins = [1, 2, 5], amount = 11
// Output: 3 
// Explanation: 11 = 5 + 5 + 1
//
// Example 2:
//
//
// Input: coins = [2], amount = 3
// Output: -1
//
//
// Note:
// You may assume that you have an infinite number of each kind of coin.
//


class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount == 0) {
            return 0;
        }
        int dp[amount + 1];
        memset(dp, -1, sizeof(dp));
        
        for (int i = 1; i <= amount; i++) {
            int minValue = -1;
            for (int j = 0; j < coins.size(); j++) {
                
                if (i - coins[j] < 0) {
                    continue;
                }
                else if (i - coins[j] == 0) {
                    minValue = 1;
                }
                else { // >0
                    if (dp[i-coins[j]] > 0) {
                        int val = dp[i-coins[j]] + 1;
                        if (minValue > 0) {
                            minValue = min(minValue, val);
                        }
                        else {
                            minValue =  val;
                        }
                    }
                }
                
                dp[i] = minValue;
            }
        }
        
        return dp[amount];
    }
};
